### PR TITLE
[PP] Allow intermediate nodes in ZB to have multiple grads

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159084

Fixes a ZB regression (https://github.com/pytorch/torchtitan/actions/runs/16478292562/job/46585646792)

Previously we only allowed an intermediate node to have 1 gradient. Recently a torchtitan ZB test started failing and I tracked to back to FusedRMSNorm grad_fn having two values `(grad, None)` (see https://github.com/pytorch/pytorch/pull/153666) and it started breaking our ZB tests.

This PR allows `stage_backward_weight` intermediate nodes to have multiple grads (it sums them together or if the grad value is None, then ignores it). Here is an example where the backward would have two grad values (gI1, gI2):

```python
class Func(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return x, 2
    @staticmethod
    def backward(ctx, gI1, gI2):
        assert gI2 is None
        return gI1
```

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

### CODE REVIEW

File: torch/distributed/pipelining/_backward.py [hunk 1]
```diff
@@ -235,11 +235,19 @@ def stage_backward_weight(
         weight_grads.append(weight.grad)
 
     for param_group in param_groups:
-        # TODO: Handle case where intermediate can have multiple outputs
```
Comment (H-Huang) [inline torch/distributed/pipelining/_backward.py:238]: This PR fixes this TODO

**Review by tianyu-l — COMMENTED** [review]
The regression came from recent support of fused RMSNorm kernels https://github.com/pytorch/pytorch/pull/153666

I know that in the past it was going through some decomposition at aten level, but I didn't pay attention to why we are seeing more fields passed around.

If this fix is a general improvement, rather than specifically tailors to this regression, then sounds good to me.

**Comment by H-Huang** [comment]
> The regression came from recent support of fused RMSNorm kernels https://github.com/pytorch/pytorch/pull/153666

Thanks for providing this context.

Yep, this fix is a general improvement, that change just happened to surface it!

**Review by tianyu-l — APPROVED** [review]
stamp to unblock

**Comment by AaronWang04** [comment]
> The regression came from recent support of fused RMSNorm kernels #153666
> 
> I know that in the past it was going through some decomposition at aten level, but I didn't pay attention to why we are seeing more fields passed around.
> 
> If this fix is a general improvement, rather than specifically tailors to this regression, then sounds good to me.

Just to add some context, an unfused composite implementation relies on autograd to save intermediate values for backprop

In a fused implementation, autograd doesn't have context as to what happens within the fused kernel, thus we have to return some intermediate values so that they can be used and do not have to be recomputed during the backward pass
